File: Java/src/pi_parallel.java
import java.util.stream.IntStream;

public class pi_parallel {

    private static int num_steps = 1000000000;
    private static double step = 1.0 / num_steps;
    private static int threads = 4;

    private static double partial_sum(Integer part) {
        double pi_sum = 0.0;
        for (int i = part; i < num_steps; i += threads) {
            double x = (i + 0.5) * step;
            pi_sum += 4 / (1 + x * x);
        }
        return pi_sum;
    }

    public static void main(String[] args) {

        double pi_sum = IntStream.range(0, threads).parallel()
                .mapToDouble(pi_parallel::partial_sum)
                .reduce(0.0, (a, b) -> a + b);

        double pi = pi_sum * step;

        System.out.println(pi);
    }
}

File: Go/pi_parallel.go
package main

const numSteps = 1000000000
const step = 1.0 / float64(numSteps)
const threads = 4

func partialSum(part int) (piSum float64) {
	for i := part; i < numSteps; i += threads {
		x := (float64(i) + 0.5) * step
		piSum += 4.0 / (1.0 + x*x)
	}
	return
}

func main() {
	res := make(chan float64)

	for i := 0; i < threads; i++ {
		go func(i int) {
			res <- partialSum(i)
		}(i)
	}

	pi := 0.0
	for i := 0; i < threads; i++ {
		pi += <-res
	}
	pi *= step

	println(pi)
}
